File: source/tools/rmgen/rectplacer.cpp
#include "stdafx.h"
#include "rectplacer.h"
#include "map.h"

RectPlacer::RectPlacer(int x1, int y1, int x2, int y2)
{
	this->x1 = x1;
	this->y1 = y1;
	this->x2 = x2;
	this->y2 = y2;
}

RectPlacer::~RectPlacer(void)
{
}

bool RectPlacer::place(Map* m, Constraint* constr, std::vector<Point>& ret) {
	for(int x=x1; x<x2; x++) {
		for(int y=y1; y<y2; y++) {
			if(m->validT(x,y) && constr->allows(m,x,y)) {
				ret.push_back(Point(x,y));
			}
			else {
				return false;
			}
		}
	}
	return true;
}

File: binaries/data/mods/public/maps/random/rmgen/placer/noncentered/RectPlacer.js
/**
 * The RectPlacer returns all tiles between the two given points that meet the Constraint.
 */
function RectPlacer(start, end, failFraction = Infinity)
{
	this.bounds = getBoundingBox([start, end]);
	this.bounds.min.floor();
	this.bounds.max.floor();
	this.failFraction = failFraction;
}

RectPlacer.prototype.place = function(constraint)
{
	const bboxPoints = getPointsInBoundingBox(this.bounds);
	const points = bboxPoints.filter(point => g_Map.inMapBounds(point) && constraint.allows(point));
	return (bboxPoints.length - points.length) / bboxPoints.length <= this.failFraction ? points : undefined;
};
